src/upgradeConfig.js

```diff
@@ -1,110 +1,91 @@
 const { presets: oldPresets, plugins: oldPlugins } = require('./packageData');
 
-// TODO: fix all of this
-function changePresets(config, options = {}) {
-  let presets = config.presets;
-
+function transformPresets(presets, options = {}) {
   if (!Array.isArray(presets) && typeof presets === 'string') {
-    presets = config.presets = config.presets.split(',').map((preset) => preset.trim());
+    presets = presets.split(',').map(preset => preset.trim());
   }
 
-  // check if presets are there
   if (presets) {
-    // assume it's an array
-    for (let i = 0; i < presets.length; i++) {
-      let preset = presets[i];
-      const presetsToReplace = Object.keys(oldPresets);
-
-      // check if it's a preset with options (an array)
-      if (Array.isArray(preset)) {
-        if (preset[0].indexOf('babel-preset') !== 0 && preset[0].indexOf('@babel/') !== 0) {
-          preset[0] = `babel-preset-${preset[0]}`;
-        }
-        if (presetsToReplace.includes(preset[0])) {
-          if (oldPresets[preset[0]]) {
-            preset[0] = oldPresets[preset[0]];
-          } else {
-            presets.splice(i, 1);
-            i--;
-          }
-        }
-      } else {
-        if (preset.indexOf('babel-preset') !== 0 && preset.indexOf('@babel/') !== 0) {
-          preset = `babel-preset-${preset}`;
-        }
-        if (presetsToReplace.includes(preset)) {
-          if (oldPresets[preset]) {
-            presets[i] = oldPresets[preset];
-          } else {
-            presets.splice(i, 1);
-            i--;
-          }
-        }
-      }
-    }
+    const newPresets = presets.map(transformPreset).filter(Boolean);
 
-    if (options.hasFlow && !presets.includes('@babel/preset-flow')) {
-      presets.push('@babel/preset-flow');
-    }
+    return options.hasFlow ? appendFlowIfNeeded(newPresets) : newPresets;
   }
+
+  return presets;
 }
 
-function changePlugins(config) {
-  let plugins = config.plugins;
+function appendFlowIfNeeded(presets) {
+  return presets.find(preset => preset === '@babel/preset-flow' || preset[0] === '@babel/preset-flow')
+    ? presets
+    : presets.concat(['@babel/preset-flow']);
+}
+
+function transformPreset(preset) {
+  let presetName = Array.isArray(preset) ? preset[0] : preset;
+  const presetOptions = Array.isArray(preset) ? preset[1] : {};
```

Comment:
normalizing presets to be "__Name" and "__Options" and then transforming them back down to a string at the end allows more code to be shared preset/plugin formats

might seem a bit unnecessary right now, but it also simplifies operations on options a bit more